picasso/src/main/java/com/squareup/picasso/SourceMarker.java

```diff
@@ -0,0 +1,185 @@
+
+  public SourceMarker(Source source) {
+    this.markSource = new MarkSource(source);
+    this.markBuffer = new Buffer();
+    this.userSource = Okio.buffer(markSource);
+    this.userBuffer = userSource.buffer();
+  }
+
+  public BufferedSource source() {
+    return userSource;
+  }
+
+  /**
+   * Marks the current position in the stream as one to potentially return back to. Returns the
+   * offset of this position. Call {@link #reset(long)} with this position to return to it later. It
+   * is an error to call {@link #reset(long)} after consuming more than {@code readLimit} bytes from
+   * {@linkplain #source() the source}.
+   */
+  public long mark(long readLimit) throws IOException {
+    if (readLimit < 0L) {
+      throw new IllegalArgumentException("readLimit < 0: " + readLimit);
+    }
+
+    if (closed) {
+      throw new IllegalStateException("closed");
+    }
+
+    // Mark the current position in the buffered source.
+    long userOffset = offset - userBuffer.size();
+
+    // If this is a new mark promote userBuffer data into the markBuffer.
+    if (mark == -1L) {
+      markBuffer.writeAll(userBuffer);
+      mark = userOffset;
+      offset = userOffset;
+    }
+
+    // Grow the limit if necessary.
+    long newMarkBufferLimit = userOffset + readLimit;
+    if (newMarkBufferLimit < 0) newMarkBufferLimit = Long.MAX_VALUE; // Long overflow!
+    limit = Math.max(limit, newMarkBufferLimit);
+
+    return userOffset;
+  }
+
+  /** Resets {@linkplain #source() the source} to {@code userOffset}. */
+  public void reset(long userOffset) throws IOException {
+    if (closed) {
+      throw new IllegalStateException("closed");
+    }
+
+    if (userOffset < mark // userOffset is before mark.
+        || userOffset > limit // userOffset is beyond limit.
+        || userOffset > mark + markBuffer.size() // userOffset is in the future.
+        || offset - userBuffer.size() > limit) { // Stream advanced beyond limit.
+      throw new IOException("cannot reset to " + userOffset + ": out of range");
+    }
+
+    // Clear userBuffer to cause data at 'offset' to be returned by the next read.
+    offset = userOffset;
+    userBuffer.clear();
+  }
+
+  final class MarkSource extends ForwardingSource {
+    MarkSource(Source source) {
+      super(source);
+    }
+
+    @Override public long read(Buffer sink, long byteCount) throws IOException {
+      if (closed) {
+        throw new IllegalStateException("closed");
+      }
+
+      // If there's no mark, go to the underlying source.
+      if (mark == -1L) {
+        long result = super.read(sink, byteCount);
+        if (result == -1L) return -1L;
+        offset += result;
+        return result;
+      }
+
+      // If we can read from markBuffer, do that.
+      if (offset < mark + markBuffer.size()) {
+        long posInBuffer = offset - mark;
+        long result = Math.min(byteCount, markBuffer.size() - posInBuffer);
+        markBuffer.copyTo(sink, posInBuffer, result);
```

Comment:
We'll only be consuming this as an `InputStream` so it makes me sad that this will be doing double copies for everything.

Reply:
Will propose something better-suited to Picasso’s use.